_drafts/firmware-update-architecture.md

```diff
@@ -0,0 +1,834 @@
+    uint32_t addr = (uint32_t)(slot == IMAGE_SLOT_1 ?
+                               &__slot1rom_start__ : &__slot2rom_start__);
+
+    uint8_t *data_ptr = (uint8_t *)hdr;
+    for (int i = 0; i < sizeof(image_hdr_t); ++i) {
+        flash_program_byte(addr + i, data_ptr[i]);
+    }
+
+    return 0;
+}
+```
+
+Just as writing the header is marks an image as valid, we can invalidate an
+image simply by zeroing it out. This is much faster than erasing the whole
+flash, and won't leave us in a half-working state (e.g. having erased only half
+of the image).
+
+```c
+int dfu_invalidate_image(image_slot_t slot) {
+    // We just write 0s over the image header
+    uint32_t addr = (uint32_t)(slot == IMAGE_SLOT_1 ?
+                               &__slot1rom_start__ : &__slot2rom_start__);
+    for (int i = 0; i < sizeof(image_hdr_t); ++i) {
+        flash_program_byte(addr + i, 0);
+    }
+
+    return 0;
+}
+```
+
+### Shared Memory
+
+It is often useful to exchange information between our different programs. The
+application may want to signal to the loader that it should go into DFU mode on
+the next reboot, and the loader may want to pass arguments to the application.
+
+On some platforms, scratch registers can be used as mailboxes between the
+different programs. This is the case for the STM32. For all others we must make
+do with RAM. "But François, isn't RAM volatile"? Not quite: as long as your MCU
+remains powered, the SRAM will hold its state. For short durations, you can
+treat is as non-volatile storage.
+
+The simplest way to carve out an area of RAM to be shared across our programs is
+to declare it in a shared linker script. We've covered linker scripts on
+Interrupt, so I won't detail the syntax again. This is what it looks like:
+
+```
+// memory_map.ld
+MEMORY
+{
+  sharedram (rwx): ORIGIN = 0x20000000, LENGTH = 256
+  ram (rwx)      : ORIGIN = 0x20000100, LENGTH = 128K - LENGTH(sharedram)
+}
+
+SECTIONS
+{
+    .shared_memory (NOLOAD) : {
+        KEEP(*(.shared_memory))
+    } >sharedram
+}
+```
+
+This shared linker script can then be included by the individual program's
+linker script with the `INCLUDE` directive.
+
+Now that we have a region of RAM defined, we can define a structure for it. You
+will want that structure to be packed to guard against alignment mismatches
+across your programs as you update compiler flags and versions. Here again we
+use the `section` attribut to assign our symbol to the `.shared_memory` section.
+
+```c
+// shared_memory.c
+typedef struct __attribute__((packed)) {
+    uint32_t magic;
+    uint32_t flags;
+} shared_memory_t;
+
+shared_memory_t shared_memory __attribute__((section(".shared_memory")));
+```
+
+In the event of power loss, RAM will lost its state. To detect those events we
+use a magic value here as well. On boot, we check the magic number and
+initialize the struct if it isn't what we'd expect.
+
+```c
+void shared_memory_init(void) {
+    if (shared_memory.magic != MAGIC) {
```

Comment:
Some missed opportunities for spelling things in ascii :p 

`const uint32_t MAGIC = 0xbadcafe;` ->

`const uint32_t MAGIC = 0x4d524853;`

```
xxd &shared_memory 4
00000000: 5348 524d                                SHRM
```